_posts/2018-01-01-testing-code-that-isnt-there-yet.md

```diff
@@ -0,0 +1,671 @@
+    const collectionName = 'hotels';
+    const hotelId = 'hotelId';
+
+    const connectedClientDouble = {
+        collection: sinon.spy()
+    };
+    const ctxDouble = {
+        params: {
+            hotelId
+        }
+    };
+
+    beforeEach(() => {
+        connectedClientDouble.collection.reset(); // (1)
+    });
+
+    it('should return a route handler', () => {
+        const routeHandler = createHotelPhotosRouteHandler(connectedClientDouble, collectionName);
+
+        expect(routeHandler).to.be.a('function');
+    });
+
+    describe('route handler', () => {
+        it('should fetch hotels collection from DB', () => {
+            const routeHandler = createHotelPhotosRouteHandler(connectedClientDouble, collectionName);
+
+            routeHandler(ctxDouble);
+
+            expect(connectedClientDouble.collection)
+                .to.have.been.calledWithExactly('hotels')
+                .to.have.been.calledOnce;
+        });
+
+        it('should return hotel photos collection', () => {
+            const routeHandler = createHotelPhotosRouteHandler(connectedClientDouble, collectionName);
+            const photosCollection = routeHandler(ctxDouble); // (2)
+
+            expect(photosCollection).to.deep.equal([
+                'photo-1.jpg', 'photo-2.jpg', 'photo-3.jpg'
+            ]);
+        });
+    });
+});
+```
+
+```javascript
+export default function createHotelPhotosRouteHandler(dbClient, collectionName) {
+    dbClient.collection(collectionName);
+
+    return () => [ 'photo-1.jpg', 'photo-2.jpg', 'photo-3.jpg' ];
+}
+```
+
+1. It is a very good practice (most of the time a requirement) to reset calls count for spies
+   if we check in tests if they were called given amount of times. Remember that you should
+   be able to call all unit tests independently, at any given moment, in any order. That being
+   said, one spy call should not affect the other call in other test. Same goes for stubs.
+
+1. At this point, we still want to have photos returned. But since this is a route handler,
+   I guess it should not return a plain array, but a response with status code and a body
+   holding that array.
+
+<blockquote>
+For the sake of readability, I will add only a single test here and there to not overwhelme
+anyone with the whole codebase.
+</blockquote>
+
+In order for Koa to return a response with given status code and a body,
+we need to set `status` and `body` properties of `response` property of `ctx` (I will explain it shortly).
+This is, again, a thing worth discovering before doing any coding.<br>
+
+### "Learn how to use your tool, before using it."
+<br>
+So, if we want to set `200` and a body with that collection of photos, we need to do something like:
+
+```javascript
+ctx.response.status = 200;
+ctx.response.body = [ 'photo-1.jpg', 'photo-2.jpg', 'photo-3.jpg' ];
+```
+
+Let's write a test for that:
+
+```javascript
+const ctxDouble = {
+    response: { // (1)
```

Comment:
are we now in a separate test file or "evolving" the previous test file? if not, ```params``` seems gone from the ```ctx``` object - I know it's returning later on in the article, but might be a bit confusing here. i guess it will be clearer once we have the repository to follow your steps (as mentioned in the PR description)

Reply:
Would you still leave the code here and have the repository as a secondary source of code?

Reply:
**Only** having the code in a repository - and not within the blog - would hamper the reading "flow" IMHO. So rather keep the relevant parts in the article and provide a separate repository to play around and follow up with (as you suggested above).

Reply:
OK, thanks.